crates/ecs/src/filter.rs

```diff
@@ -0,0 +1,345 @@
+///     println!("A non-player entity is at position {:?}.", position);
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Not<T: SystemParameter> {
+    phantom: PhantomData<T>,
+}
+
+impl<T: SystemParameter> SystemParameter for Not<T> {
+    type BorrowedData<'components> = <T as SystemParameter>::BorrowedData<'components>;
+
+    fn borrow(world: &World) -> Self::BorrowedData<'_> {
+        <T as SystemParameter>::borrow(world)
+    }
+
+    unsafe fn fetch_parameter(
+        borrowed: &mut Self::BorrowedData<'_>,
+        entity: Entity,
+    ) -> Option<Self> {
+        if <T as SystemParameter>::fetch_parameter(borrowed, entity).is_some() {
+            None
+        } else {
+            Some(Self {
+                phantom: PhantomData::default(),
+            })
+        }
+    }
+
+    fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+        <T as SystemParameter>::component_accesses()
+    }
+}
+
+/// A query filter that matches any entity.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::Any;
+/// fn all_entities(_: Any) {
+///     println!("Hello from an entity!");
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Any {}
+
+impl SystemParameter for Any {
+    type BorrowedData<'components> = ();
+
+    fn borrow(_world: &World) -> Self::BorrowedData<'_> {}
+
+    unsafe fn fetch_parameter(
+        _borrowed: &mut Self::BorrowedData<'_>,
+        _entity: Entity,
+    ) -> Option<Self> {
+        Some(Self {})
+    }
+
+    fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+        vec![]
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{Read, Write};
+    use proptest::proptest;
+    use test_log::test;
+
+    #[derive(Debug)]
+    struct A;
+
+    #[derive(Debug)]
+    struct B;
+
+    #[derive(Debug)]
+    struct C;
+
+    fn test_filter<Filter: SystemParameter>((a, b, c): (bool, bool, bool)) -> Option<Filter> {
+        let mut world: World = Default::default();
+
+        let entity: Entity = Default::default();
+        world.entities.push(entity);
+
+        if a {
+            world.create_component_vec_and_add(entity, A);
+        }
+        if b {
+            world.create_component_vec_and_add(entity, B);
+        }
+        if c {
+            world.create_component_vec_and_add(entity, C);
+        }
+
+        let mut borrowed = <Filter as SystemParameter>::borrow(&world);
+        unsafe { <Filter as SystemParameter>::fetch_parameter(&mut borrowed, entity) }
+    }
+
+    macro_rules! identity {
+        ($vars:expr, $expr:ty, $expected:literal) => {
+            assert_eq!(test_filter::<$expr>($vars).is_some(), $expected);
+        };
+        ($vars:expr, $lhs:ty, $rhs:ty) => {
+            assert_eq!(
+                test_filter::<$lhs>($vars).is_some(),
+                test_filter::<$rhs>($vars).is_some(),
+            );
+        };
+    }
+
+    proptest! {
+        #[test]
+        fn test_filter_identities(vars: (bool, bool, bool)) {
```

Comment:
Could you split this up into several tests instead of one single large one? Ideally, you want a test to only have one reason to fail because that makes it easier to see what's not working. 

Right now if this test were to fail, you'd have to carefully look at the line numbers to see which of the actual assertions are failing. And the later assertions wouldn't even run, so you couldn't know if they still pass or not.

Reply:
(also, is there a better name than `vars`? That name doesn't say a whole lot about what it's contents are)